### PR TITLE
Sequence execution

### PR DESCRIPTION
**Pull request by ioana-blue**
The bulk of the work is done. Submitting for review to get some feedback.

### CODE REVIEW

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 1]
```diff
@@ -722,6 +735,251 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
...
+                        val seqActivation = makeSeqActivation(user, action, seqActivationId, wskActivations, topmost, wskActivations.last.response, cause)
+                        (seqActivationId, Some(seqActivation))
+                    }
+                } else {
+                    // non-blocking sequence execution, return activation id
+                    Future.successful((seqActivationId, None))
+                }
+            } else {
+                // not topmost, blocking for sure, wait for the "aggregate" future to finish (no timeout)
+                val seqResult = Await.ready(futureSeqResult, Duration.Inf).value.get
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: is there a way to avoid using Await? this blocks the thread and is not guaranteed to be deadlock free. also the Duration.Inf is worrisome.

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 2]
```diff
@@ -722,6 +735,251 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
...
+            cause: Option[ActivationId])(implicit transid: TransactionId): WhiskActivation = {
+        // at this point all futures are finished, filter out the successful ones
+        val failedIdx = futureWskActivations.indexWhere {
+            _.value match {
+                case Some(Failure(_)) => true
+                case _ => false
+            }
+        }
+        // all futures are done, this should return immediately
+        val successfulActivations = Await.result(Future.sequence(futureWskActivations.slice(0, failedIdx)), Duration.Inf)
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: See comment about await earlier - your comment suggests all futures are done so why use await?

**Comment by rabbah** [comment]
did you consider moving all the sequence scheduling code to its own trait? (which then you can test for all the corner cases perhaps).

re this: >  remove old sequence code (and fix tests that check for pipe.js specific things)
can we have a feature toggle so that we can back out the switch if needed and then when we’re sure cleanup in a new pr.

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 3]
```diff
@@ -781,6 +796,318 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
...
+                val activation = seqActivation getOrElse {
+                    val wskActivationEithers = wskActivationTuples.map(_._1)
+                    val end = Instant.now(Clock.systemUTC())
+                    // the response of the sequence is the response of the very last activation
+                    makeSequenceActivation(user, action, seqActivationId, wskActivationEithers, topmost, cause, start, end)
+                }
+                storeSequenceActivation(activation)
+            case Failure(t: Throwable) =>
+                // consider this whisk error
+                error(this, s"Sequence activation 'seqActivationId' failed: ${t.getMessage}")
```
Comment (psuter) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: missing `$` before `seqActivationId`

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 4]
```diff
@@ -781,6 +796,318 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
...
+            user: Identity,
+            action: WhiskAction,
+            activationId: ActivationId,
+            wskActivationEithers: Vector[Either[ActivationResponse, WhiskActivation]],
+            topmost: Boolean,
+            cause: Option[ActivationId],
+            start: Instant,
+            end: Instant) : WhiskActivation = {
+        // extract all successful activations from the vector of activation eithers
+        val leftIndex = wskActivationEithers.indexWhere {
```
Comment (psuter) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: `....indexWhere(_.isLeft)`

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 5]
```diff
@@ -781,6 +796,318 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
...
+            blocking: Boolean,
+            topmost: Boolean,
+            components: Vector[FullyQualifiedEntityName],
+            cause: Option[ActivationId],
+            atomicActionsCount: Int)(implicit transid: TransactionId): Future[(ActivationId, Option[WhiskActivation], Int)] = {
+        // create new activation id that corresponds to the sequence
+        val seqActivationId = activationIdFactory.make()
+        info(this, s"Invoking sequence $action topmost $topmost activationid '$seqActivationId'")
+        val start = Instant.now(Clock.systemUTC())
+        var seqActivation: Option[WhiskActivation] = None
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: this `var` is potentially dangerous as it's racy wrt the future below that assigns it.

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 6]
```diff
@@ -781,6 +796,318 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
...
+            if (topmost) { // need to deal with blocking and closing connection
+                if (blocking) {
+                    val timeout = maxWaitForBlockingActivation + blockingInvokeGrace
+                    val futureSeqResultTimeout = futureSeqResult withTimeout(timeout, new BlockingInvokeTimeout(seqActivationId))
+                    // if the future fails with a timeout, the failure is dealt with at the caller level
+                    futureSeqResultTimeout map { wskActivationTuples =>
+                        val wskActivationEithers = wskActivationTuples.map(_._1)
+                        // the execution of the sequence was successful, return the result
+                        val end = Instant.now(Clock.systemUTC())
+                        seqActivation = Some(makeSequenceActivation(user, action, seqActivationId, wskActivationEithers, topmost, cause, start, end))
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: this flow for creating the seqActivation is repeated three times - the only difference is the `end` marker. 
since the `end` marker is not that important for the sequence, is it really worth the code duplication?

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 7]
```diff
@@ -781,6 +796,318 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
...
+        // the execution can reach here without a seqActivation due to non-blocking activations OR blocking activations that reach the blocking invoke timeout
+        // futureSeqResult should always be successful, if failed, there is an error
+        futureSeqResult onComplete {
+            case Success(wskActivationTuples) =>
+                // all activations were successful
+                val activation = seqActivation getOrElse {
+                    val wskActivationEithers = wskActivationTuples.map(_._1)
+                    val end = Instant.now(Clock.systemUTC())
+                    // the response of the sequence is the response of the very last activation
+                    makeSequenceActivation(user, action, seqActivationId, wskActivationEithers, topmost, cause, start, end)
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: this can throw an exception in which case no activation will be created - either this should be wrapped with a future itself or a try to allow for recovery.

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 8]
```diff
@@ -781,6 +796,318 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
...
+                    // non-blocking sequence execution, return activation id
+                    Future.successful((seqActivationId, None, 0))
+                }
+            } else {
+                // not topmost, no need to worry about terminating incoming request
+                futureSeqResult map { wskActivationTuples =>
+                     val wskActivationEithers = wskActivationTuples.map(_._1)
+                    // all activations are successful, the result of the sequence is the result of the last activation
+                    val end = Instant.now(Clock.systemUTC())
+                    seqActivation = Some(makeSequenceActivation(user, action, seqActivationId, wskActivationEithers, topmost, cause, start, end))
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: makeSequenceActivation could throw an exception and this will cause the future to complete with an error and not create an activation.

File: tests/src/system/basic/WskSequenceTests.scala [hunk 9]
```diff
@@ -68,23 +79,359 @@ class WskSequenceTests
...
+                    checkSequenceLogsAndAnnotations(activation, 3) // 3 activations in this sequence
+                    activation.cause shouldBe None   // topmost sequence
+                    val result = activation.response.result.get
+                    result.fields.get("payload") shouldBe defined
+                    result.fields.get("length") should not be defined
+                    result.fields.get("lines") shouldBe Some(JsArray(Vector(now.toJson)))
+            }
+    }
+
+    it should "create and run a sequence in a package with parameters" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/system/basic/WskSequenceTests.scala]: need a test where package is private/owned by other namespace.

**Comment by rabbah** [comment]
closing for #1533.
